### ISSUE TITLE
Make Service Bus SDK class properties virtual

### ISSUE THREAD

**Issue by NateB2** [Azure/azure-sdk-for-net#17602]
**Library or service name.**
Azure.Messaging.ServiceBus
 
**Is your feature request related to a problem? Please describe.**
We are trying to migrate to the new Azure.Messaging.ServiceBus library, and are having issues mocking some of the classes - namely ServiceBusSender and ServiceBusReceiver.  Our mocking framework, FakeItEasy, refuses to mock the "IsClosed" property on those classes because it isn't virtual. Ideally, it would be great if the class properties as well as the methods were virtual, which would allow those classes to be fully mocked and would fix the issue we're currently experiencing.

**Comment by jsquire** [comment]
Hi @NateB2.  Thank you for your feedback and we regret that you're experiencing difficulties.   We definitely aim to be fully testable and support injection or mocking for public members.  Generally, though, we try to avoid virtual properties due to the performance overhead and prefer to allow values to be injected at construction for read-only members.

In the case of `IsClosed`, the value can be set by simply calling the `CloseAsync` method.  Because there is no active connection to the service, the call will no-op and set `IsClosed` to true.    I can't claim knowledge of FakeItEasy, but using Moq it would look like:

```csharp
var mockSender = new Mock<ServiceBusSender>();	
await mockSender.Object.CloseAsync();
```

When taking a look at this, I did notice that we do have two property gaps for the sender -  `FullyQualifiedNamespace` and `EntityPath`  and are likely to have the same gaps on the receiver, as they're normally constructed by the client.   We'll take a look at getting those fixed up.

**Comment by jsquire** [comment]
//cc: @JoshLove-msft

**Comment by NateB2** [comment]
[I tried the equivalent in FakeItEasy,](https://fakeiteasy.readthedocs.io/en/stable/calling-base-methods/) and received a NullReferenceException on the last line in this snippet:
```
var sender = A.Fake<ServiceBusSender>();
A.CallTo(() => sender.CloseAsync(A<CancellationToken>._)).CallsBaseMethod();
await sender.CloseAsync();
```

The call stack seems to indicate that the error is occurring in the Service Bus SDK.

**Comment by JoshLove-msft** [comment]
> [I tried the equivalent in FakeItEasy,](https://fakeiteasy.readthedocs.io/en/stable/calling-base-methods/) and received a NullReferenceException on the last line in this snippet:
> 
> ```
> var sender = A.Fake<ServiceBusSender>();
> A.CallTo(() => sender.CloseAsync(A<CancellationToken>._)).CallsBaseMethod();
> await sender.CloseAsync();
> ```
> 
> The call stack seems to indicate that the error is occurring in the Service Bus SDK.

It looks like this would suffer from the same issue as FullyQualifiedNamespace and EntityPath where the necessary dependencies are not configured as part of the ctor, but rather as part of the CreateSender factory methods.

**Comment by jsquire** [comment]
Apologies for the confusion; the oversight is on me.  

I forgot to configure Moq to pass through the call to the base and didn't account for changes in the close logic since I had last looked.  `CloseAsync` isn't taking a defensive approach to an internal member not being populated.   We'll need to make an adjustment here as well or provide another means to set the value.

**Comment by jsquire** [comment]
Fixed by #18191

**Comment by jsquire** [comment]
These changes have been merged and will be included in the next release.   Closing this out.
